src/Swarm/Game/Scenario/Status.hs

```diff
@@ -0,0 +1,235 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Swarm.Game.Scenario.Status where
+
+import Control.Arrow ((&&&))
+import Control.Lens hiding (from, (<.>))
+import Data.Aeson (
+  genericParseJSON,
+  genericToEncoding,
+  genericToJSON,
+ )
+import Data.Function (on)
+import Data.List.NonEmpty (NonEmpty)
+import Data.List.NonEmpty qualified as NE
+import Data.Map (Map)
+import Data.Map qualified as M
+import Data.Maybe (mapMaybe)
+import Data.Text (Text)
+import Data.Time (ZonedTime, diffUTCTime, zonedTimeToUTC)
+import Data.Yaml as Y
+import GHC.Generics (Generic)
+import Swarm.Game.Scenario
+import Swarm.Game.Scenario.Scoring.CodeSize
+import Swarm.Game.Scenario.Scoring.Metrics
+import Swarm.Game.Scenario.Scoring.Progress
+
+-- * Some orphan ZonedTime instances
+
+instance Eq ZonedTime where
+  (==) = (==) `on` zonedTimeToUTC
+
+instance Ord ZonedTime where
+  (<=) = (<=) `on` zonedTimeToUTC
+
+-- * High scores by various criteria
+
+data BestByCriteria
+  = BestByTime
+  | BestByTicks
+  | BestByCharCount
+  | BestByAstSize
+  deriving (Eq, Ord, Bounded, Enum, Show)
+
+describeCriteria :: BestByCriteria -> Text
+describeCriteria = \case
+  BestByTime -> "time"
+  BestByTicks -> "ticks"
+  BestByCharCount -> "char count"
+  BestByAstSize -> "AST size"
+
+data ProgressStats = ProgressStats
+  { _scenarioStarted :: ZonedTime
+  -- ^ Time when the scenario was started including time zone.
+  , _scenarioAttemptMetrics :: AttemptMetrics
+  }
+  deriving (Eq, Ord, Show, Read, Generic)
+
+makeLenses ''ProgressStats
+
+instance FromJSON ProgressStats where
+  parseJSON = genericParseJSON scenarioOptions
+
+instance ToJSON ProgressStats where
+  toEncoding = genericToEncoding scenarioOptions
+  toJSON = genericToJSON scenarioOptions
+
+type ProgressMetric = Metric ProgressStats
+
+-- * High scores by various criteria
+
+data BestRecords = BestRecords
+  { _scenarioBestByTime :: ProgressMetric
+  , _scenarioBestByTicks :: ProgressMetric
+  , _scenarioBestByCharCount :: ProgressMetric
+  , _scenarioBestByAstSize :: ProgressMetric
+  }
```

Comment:
Uh, can we have a `Map BestByCriteria ProgressStats` instead?

We can wait to reach the magic number 7, but this is already more than I would be comfortable touching.

Reply:
Created #1132